Update README.adoc (#50)
cloudbees-io/kaniko: README.adoc

----

> . The Docker config file must be formatted in JSON.

Reviewer: Reviewing the README as a whole, I realized that this second prerequisite is not needed actually. The OCI credentials configuration action mentioned in the first prerequisite takes care of creating the docker config file. Instead of the second prereq, we may want to tell the user that they need to setup the property `DOCKER_REGISTRY` and secret properties `DOCKER_USERNAME` and `DOCKER_PASSWORD`.

Author: @urvashisingh I think because we link to the OC credentials action, we do not need to explain in detail?

Reviewer: yes, that makes sense. The OC action should be explaining this in detail.

| Required?
| Description

| `destination`
| String
| Yes
| The locations of the target images to be published.
Formatted as a comma-separated list for passing multiple images.

| `build-args`
| String
| No
| The build arguments to be passed to the Kaniko build.
Formatted as a comma-separated list for passing multiple build arguments.

| `context`
| String
| No
| The path to the build context.
Default is `${{ cloudbees.workspace }}`.

| `dockerfile`
| String
| No

> | The path to the Dockerfile to be built.

Reviewer: ```suggestion
| The path to the Dockerfile.
```

Dockerfile must exist, it is not built here

| `artifact-ids`
| JSON string
| The unique identifiers for each of the published image locations (`destination`) reported to the CloudBees platform, in JSON format.
|===

== Usage examples

=== Basic example

The following is a basic usage example for this action:

[source,yaml]
----
      - name: Build a container image with Kaniko
        uses: cloudbees-io/kaniko@v1
        with:
          destination: path/to/registry/host/my-image:1.0.1,path/to/registry/host/my-image:latest
----

=== Using optional inputs

The following example specifies optional inputs:

[source,yaml]
----
      - name: Kaniko build with optional inputs
        uses: cloudbees-io/kaniko@v1
        with:
          destination: path/to/registry/host/my-image:1.0.1,path/to/registry/host/my-image:latest
          build-args: BUILDKIT_CONTEXT_KEEP_GIT_DIR=1,BUILDKIT_INLINE_CACHE=1
          context: .
          dockerfile: path/to/Dockerfile
          labels: maintainer=John Smith,version=1.0.1
          tar-path: path/to/image.tar
          verbosity: warn

----

=== Using the action output

Access the `artifact-ids` values in downstream steps using the `outputs` xref:dsl-syntax:contexts.adoc[context].

> The following is the JSON format for the `artifactIds` ouput, where `<destination>` is the specified `destination` input parameter value, and `<artifact-version-id>` is the unique identifier of the artifact version.

Reviewer: ```suggestion
The following is the JSON format for the `artifact-ids` ouput, where `<destination>` is the specified `destination` input parameter value, and `<artifact-version-id>` is the unique identifier of the artifact version.
```

[source,json,role="novalidate"]
----
{
  "<destination>": "<artifact-version-id>"
}
----

The following is an example of an `artifact-ids` JSON for two artifact IDs:

[source,json,role="novalidate"]
----
{
  "index.docker.io/example/my-docker:1.0.87": "1234abcd-56ef-gh78-9012-ijklmnop3456"
  "index.docker.io/example/my-docker:1.0.87-dev": "ab34cd12-78gh-56ef-ij78-3456mnopkl90"
}
----

Use the `artifact-ids` output as follows, where <action_step_ID> is the action step ID, and <destination_URL> is the destination URL:

> * `${{ steps.<action_step_ID>.outputs.artifactIds }}` for a JSON string of all outputted artifact ID values.
> 
> * `${{ fromJSON(steps.<action_step_ID>.outputs.artifactIds).<destination_URL> }}` for a single artifact ID value.

Reviewer: ```suggestion
* `${{ steps.<action_step_ID>.outputs.artifact-ids }}` for a JSON string of all outputted artifact ID values.

* `${{ fromJSON(steps.<action_step_ID>.outputs.artifact-ids).<destination_URL> }}` for a single artifact ID value.
```

          context: .
          dockerfile: path/to/Dockerfile
          labels: maintainer=John Smith,version=1.0.1
          tar-path: path/to/image.tar
          verbosity: warn

----

=== Using the action output

Access the `artifactIds` values in downstream steps using the `outputs` xref:dsl-syntax:contexts.adoc[context].

The following is the JSON format for the `artifactIds` ouput, where `<destination>` is the specified `destination` input parameter value, and `<artifact-version-id>` is the unique identifier of the artifact version.

[source,json,role="novalidate"]
----
{
  "<destination>": "<artifact-version-id>"
}
----

The following is an example of an `artifactIds` JSON for two artifact IDs:

[source,json,role="novalidate"]
----
{
  "index.docker.io/example/my-docker:1.0.87": "1234abcd-56ef-gh78-9012-ijklmnop3456"
  "index.docker.io/example/my-docker:1.0.87-dev": "ab34cd12-78gh-56ef-ij78-3456mnopkl90"
}
----

Use the `artifactIds` output as follows, where <action_step_ID> is the action step ID, and <destination_URL> is the destination URL:

* `${{ steps.<action_step_ID>.outputs.artifactIds }}` for a JSON string of all outputted artifact ID values.

> * `${{ fromJSON(steps.<action_step_ID>.outputs.artifactIds).<artifact_version> }}` for a single artifact ID value.

Reviewer: ```suggestion
* `${{ fromJSON(steps.<action_step_ID>.outputs.artifactIds)['<destination_URL>'] }}` for a single artifact ID value.
```

| `artifactIds`
| String
| The unique identifier of the artifact reported to the CloudBees platform.
|===

> === Using the action output

Reviewer: IMO "Using the action output" section should go under the "Usage examples" section possibly after the "Basic example" or even at the end of the examples section.

Access the `artifactIds` values in downstream steps using the `outputs` xref:dsl-syntax:contexts.adoc[context].

> To use the following examples, replace <ACTION_STEP_NAME> with the action step name, and <DESTINATION_URL> with the destination URL.

Reviewer: ```suggestion
To use the following examples, replace <STEP_ID> with the step id, and <DESTINATION_URL> with the target image location (`destination`) URL.
```

* For all artifact ID values in JSON format provided by the action, use: `${{ steps.<ACTION_STEP_NAME>.outputs.artifactIds }}`.

> * For a specific artifact ID value use: `${{ fromJSON(steps.<ACTION_STEP_NAME>.outputs.artifactIds).<DESTINATION_URL>}}`.

Reviewer: ```suggestion
* For a specific artifact ID value use: `${{ fromJSON(steps.<STEP_ID>.outputs.artifactIds)['<DESTINATION_URL>']}}`.
```

== Usage examples

=== Basic example

The following is a basic usage example for this action:

[source,yaml]
----
      - name: Build a container image with Kaniko
        uses: cloudbees-io/kaniko@v1
        with:
          destination: path/to/registry/host/my-image:1.0.1,path/to/registry/host/my-image:latest
----

=== Using optional inputs

The following example specifies optional inputs:

[source,yaml]
----
      - name: Kaniko build with optional inputs
        uses: cloudbees-io/kaniko@v1
        with:
          destination: path/to/registry/host/my-image:1.0.1,path/to/registry/host/my-image:latest
          build-args: BUILDKIT_CONTEXT_KEEP_GIT_DIR=1,BUILDKIT_INLINE_CACHE=1
          context: .
          dockerfile: path/to/Dockerfile
          labels: maintainer=John Smith,version=1.0.1
          tar-path: path/to/image.tar
          verbosity: warn

----

=== Full workflow example

The following workflow example:

* Checks out source code from a repository.
* Configures Docker credentials.
* Builds and publishes a container image with Kaniko.
* Prints the artifact IDs for both static and dynamically created destinations.
* Promotes the image to AWS ECR.

> * Accesses the `artifactIds` value in the step to promote the image to ECR.

Reviewer: The promotion step does not use the artifactIds so we should remove it from here. In future, when we enhance the `publish-deployed-artifact` action to accept artifactId as input, we could expand the example then. For now, lets remove it.
And so also, remove the step "Promote an image in ECR" from the workflow DSL below.
